File: src/main/java/net/shadowfacts/craftingslabs/block/BlockCraftingslab.java
package net.shadowfacts.craftingslabs.block;

import net.minecraft.block.BlockSlab;
import net.minecraft.block.material.Material;
import net.minecraft.block.properties.IProperty;
import net.minecraft.block.properties.PropertyBool;
import net.minecraft.block.state.BlockState;
import net.minecraft.block.state.IBlockState;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraft.world.World;
import net.shadowfacts.craftingslabs.CraftingSlabs;
import net.shadowfacts.craftingslabs.gui.GUIs;

import java.util.Random;

/**
 * @author shadowfacts
 */
public class BlockCraftingSlab extends BlockSlab {

	private static final PropertyBool VARIANT = PropertyBool.create("variant");

	public BlockCraftingSlab() {
		super(Material.wood);
		setUnlocalizedName("craftingSlab");
		useNeighborBrightness = true;
		setCreativeTab(CreativeTabs.tabDecorations);

		setDefaultState(blockState.getBaseState()
						.withProperty(VARIANT, false)
						.withProperty(HALF, EnumBlockHalf.BOTTOM));
	}

	@Override
	public String getUnlocalizedName(int meta) {
		return getUnlocalizedName();
	}

	@Override
	public boolean isDouble() {
		return false;
	}

	@Override
	public IProperty<?> getVariantProperty() {
		return VARIANT;
	}

	@Override
	public Object getVariant(ItemStack stack) {
		return false;
	}

	@Override
	public IBlockState getStateFromMeta(int meta) {
		EnumBlockHalf half = (meta & 8) != 0 ? EnumBlockHalf.TOP : EnumBlockHalf.BOTTOM;
		return getDefaultState()
				.withProperty(VARIANT, false)
				.withProperty(HALF, half);
	}

	@Override
	public int getMetaFromState(IBlockState state) {
		EnumBlockHalf half = state.getValue(HALF);
		return half == EnumBlockHalf.TOP ? 8 : 0;
	}

	@Override
	protected BlockState createBlockState() {
		return new BlockState(this, VARIANT, HALF);
	}

	@Override
	public int damageDropped(IBlockState state) {
		return 0;
	}

	@Override
	public Item getItemDropped(IBlockState state, Random rand, int fortune) {
		return CraftingSlabs.items.craftingSlab;
	}

	@Override
	public Item getItem(World worldIn, BlockPos pos) {
		return CraftingSlabs.items.craftingSlab;
	}

	@Override
	public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumFacing side, float hitX, float hitY, float hitZ) {
		player.openGui(CraftingSlabs.instance, GUIs.CRAFTING.ordinal(), world, pos.getX(), pos.getY(), pos.getZ());
		return true;
	}
}

File: src/main/kotlin/net/shadowfacts/craftingslabs/block/BlockCraftingSlab.kt
package net.shadowfacts.craftingslabs.block

import mcmultipart.api.multipart.IMultipart
import mcmultipart.api.slot.EnumFaceSlot
import mcmultipart.api.slot.IPartSlot
import net.minecraft.block.BlockSlab
import net.minecraft.block.material.Material
import net.minecraft.block.properties.IProperty
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.shadowfacts.craftingslabs.property.PropertyDummy

/**
 * @author shadowfacts
 */
class BlockCraftingSlab: BlockSlab(Material.WOOD), IMultipart {

	companion object {
		val DUMMY = PropertyDummy("dummy")
	}

	init {
		setRegistryName("crafting_slab")
		unlocalizedName = registryName.toString()
		useNeighborBrightness = true
		setCreativeTab(CreativeTabs.DECORATIONS)

		defaultState = blockState.baseState.withProperty(DUMMY, PropertyDummy.Dummy).withProperty(HALF, EnumBlockHalf.BOTTOM)
	}

	override fun getTypeForItem(stack: ItemStack?): Comparable<*> {
		return PropertyDummy.Dummy
	}

	override fun isDouble(): Boolean {
		return false
	}

	override fun getUnlocalizedName(meta: Int): String {
		return unlocalizedName
	}

	override fun getVariantProperty(): IProperty<*> {
		return DUMMY
	}

	override fun createBlockState(): BlockStateContainer {
		return BlockStateContainer(this, HALF, DUMMY)
	}

	override fun getMetaFromState(state: IBlockState): Int {
		return if (state.getValue(HALF) == EnumBlockHalf.TOP) 1 else 0
	}

	@Deprecated("")
	override fun getStateFromMeta(meta: Int): IBlockState {
		return defaultState.withProperty(HALF, if (meta == 1) EnumBlockHalf.TOP else EnumBlockHalf.BOTTOM)
	}

	private fun getHalfForPlacement(facing: EnumFacing, hitY: Float): EnumBlockHalf {
		return when (facing) {
			EnumFacing.UP -> EnumBlockHalf.TOP
			EnumFacing.DOWN -> EnumBlockHalf.BOTTOM
			else -> if (hitY > 0.5) EnumBlockHalf.TOP else EnumBlockHalf.BOTTOM
		}
	}

	override fun getStateForPlacement(world: World, pos: BlockPos, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float, meta: Int, placer: EntityLivingBase, hand: EnumHand): IBlockState {
		return defaultState.withProperty(HALF, getHalfForPlacement(facing, hitY))
	}

	override fun getSlotForPlacement(world: World, pos: BlockPos, state: IBlockState, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float, placer: EntityLivingBase): IPartSlot {
		return when (getHalfForPlacement(facing, hitY)) {
			EnumBlockHalf.TOP -> EnumFaceSlot.UP
			EnumBlockHalf.BOTTOM -> EnumFaceSlot.DOWN
		}
	}

	override fun getSlotFromWorld(world: IBlockAccess, pos: BlockPos, state: IBlockState): IPartSlot {
		return when (state.getValue(HALF)!!) {
			EnumBlockHalf.TOP -> EnumFaceSlot.UP
			EnumBlockHalf.BOTTOM -> EnumFaceSlot.DOWN
		}
	}

}